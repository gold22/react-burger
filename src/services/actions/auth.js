export const REGISTER_USER_REQUEST = 'REGISTER_USER_REQUEST';
export const REGISTER_USER_SUCCESS = 'REGISTER_USER_SUCCESS';
export const REGISTER_USER_ERROR = 'REGISTER_USER_ERROR';

export const registerUser = (user) => async (dispatch, getState, apiClient) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result.user,
            accessToken: result.accessToken,
            refreshToken: result.refreshToken,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: error.message,
        });
    }
};
