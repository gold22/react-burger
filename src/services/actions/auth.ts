import {
    REGISTER_USER_REQUEST,
    REGISTER_USER_SUCCESS,
    REGISTER_USER_ERROR,
    LOGIN_USER_REQUEST,
    LOGIN_USER_SUCCESS,
    LOGIN_USER_ERROR,
    LOGOUT_USER_REQUEST,
    LOGOUT_USER_SUCCESS,
    LOGOUT_USER_ERROR,
    GET_USER_REQUEST,
    GET_USER_SUCCESS,
    GET_USER_ERROR,
    UPDATE_USER_REQUEST,
    UPDATE_USER_SUCCESS,
    UPDATE_USER_ERROR,
} from '../constants/auth';
import { TApiUser, TApiUserCredentials } from '../types/api';
import { TDispatch, TGetState, TThunk } from '../types/store';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

export const registerUser: TThunk = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logInUser: TThunk = (user: TApiUserCredentials) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGIN_USER_REQUEST,
    });
    try {
        const result = await apiClient.logInUser(user);
        dispatch({
            type: LOGIN_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: LOGIN_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logOutUser: TThunk = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGOUT_USER_REQUEST,
    });
    try {
        await apiClient.logOutUser();
        dispatch({
            type: LOGOUT_USER_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: LOGOUT_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const getUser = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: GET_USER_REQUEST,
    });
    try {
        const user = await apiClient.getUser();
        dispatch({
            type: GET_USER_SUCCESS,
            user,
        });
    } catch (error) {
        dispatch({
            type: GET_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const updateUser = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: UPDATE_USER_REQUEST,
    });
    try {
        const result = await apiClient.updateUser(user);
        dispatch({
            type: UPDATE_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: UPDATE_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};
