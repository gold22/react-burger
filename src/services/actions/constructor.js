import { v4 as uuid } from 'uuid';
import {
    SET_BUN,
    ADD_INGREDIENT,
    REMOVE_INGREDIENT,
    REMOVE_INGREDIENTS,
    MOVE_INGREDIENT,
} from '../constants/constructor';

export const setBun = (bun) => ({
    type: SET_BUN,
    bun,
});

export const addIngredient = (ingredient, index) => ({
    type: ADD_INGREDIENT,
    ingredient: {
        ...ingredient,
        uuid: uuid(),
    },
    index,
});

export const removeIngredient = (index) => ({
    type: REMOVE_INGREDIENT,
    index,
});

export const removeIngredients = () => ({
    type: REMOVE_INGREDIENTS,
});

export const moveIngredient = (oldIndex, newIndex) => ({
    type: MOVE_INGREDIENT,
    oldIndex,
    newIndex,
});
