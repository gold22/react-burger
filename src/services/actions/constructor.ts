import { v4 as uuid } from 'uuid';
import {
    SET_BUN,
    ADD_INGREDIENT,
    REMOVE_INGREDIENT,
    REMOVE_INGREDIENTS,
    MOVE_INGREDIENT,
} from '../constants/constructor';
import { TIngredient } from '../types';
import {
    IAddIngredientAction,
    IMoveIngredientAction,
    IRemoveIngredientAction,
    IRemoveIngredientsAction,
    ISetBunAction,
} from '../types/actions/constructor';

export const setBun = (bun: TIngredient): ISetBunAction => ({
    type: SET_BUN,
    bun,
});

export const addIngredient = (ingredient: TIngredient, index: number): IAddIngredientAction => ({
    type: ADD_INGREDIENT,
    ingredient: {
        ...ingredient,
        uuid: uuid(),
    },
    index,
});

export const removeIngredient = (index: number): IRemoveIngredientAction => ({
    type: REMOVE_INGREDIENT,
    index,
});

export const removeIngredients = (): IRemoveIngredientsAction => ({
    type: REMOVE_INGREDIENTS,
});

export const moveIngredient = (oldIndex: number, newIndex: number): IMoveIngredientAction => ({
    type: MOVE_INGREDIENT,
    oldIndex,
    newIndex,
});
