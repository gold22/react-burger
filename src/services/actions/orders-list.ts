import {
    ORDERS_LIST_CONNECTION_OPEN,
    ORDERS_LIST_CONNECTION_CLOSE,
} from '../constants/orders-list';
import {
    IOrdersListConnectionCloseAction,
    IOrdersListConnectionOpenAction,
} from '../types/actions/orders-list';

export const openOrdersListConnection = (): IOrdersListConnectionOpenAction => ({
    type: ORDERS_LIST_CONNECTION_OPEN,
});

export const closeOrdersListConnection = (): IOrdersListConnectionCloseAction => ({
    type: ORDERS_LIST_CONNECTION_CLOSE,
});
