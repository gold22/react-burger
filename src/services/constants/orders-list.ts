export const ORDERS_LIST_CONNECTION_OPEN: 'ORDERS_LIST_CONNECTION_OPEN' = 'ORDERS_LIST_CONNECTION_OPEN';
export const ORDERS_LIST_CONNECTION_CLOSE: 'ORDERS_LIST_CONNECTION_CLOSE' = 'ORDERS_LIST_CONNECTION_CLOSE';
export const ORDERS_LIST_CONNECTION_SUCCESS: 'ORDERS_LIST_CONNECTION_SUCCESS' = 'ORDERS_LIST_CONNECTION_SUCCESS';
export const ORDERS_LIST_CONNECTION_ERROR: 'ORDERS_LIST_CONNECTION_ERROR' = 'ORDERS_LIST_CONNECTION_ERROR';
export const ORDERS_LIST_CONNECTION_CLOSED: 'ORDERS_LIST_CONNECTION_CLOSED' = 'ORDERS_LIST_CONNECTION_CLOSED';
export const ORDERS_LIST_RECEIVED: 'ORDERS_LIST_RECEIVED' = 'ORDERS_LIST_RECEIVED';
