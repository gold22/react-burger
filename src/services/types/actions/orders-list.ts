import {
    ORDERS_LIST_CONNECTION_OPEN,
    ORDERS_LIST_CONNECTION_CLOSE,
    ORDERS_LIST_CONNECTION_SUCCESS,
    ORDERS_LIST_CONNECTION_ERROR,
    ORDERS_LIST_CONNECTION_CLOSED,
    ORDERS_LIST_RECEIVED,
} from '../../constants/orders-list';
import { TApiOrders } from '../api';

export interface IOrdersListConnectionOpenAction {
    readonly type: typeof ORDERS_LIST_CONNECTION_OPEN;
}

export interface IOrdersListConnectionCloseAction {
    readonly type: typeof ORDERS_LIST_CONNECTION_CLOSE;
}

export interface IOrdersListConnectionSuccessAction {
    readonly type: typeof ORDERS_LIST_CONNECTION_SUCCESS;
}

export interface IOrdersListConnectionFailedAction {
    readonly type: typeof ORDERS_LIST_CONNECTION_ERROR;
    readonly message: string | null;
}

export interface IOrdersListConnectionClosedAction {
    readonly type: typeof ORDERS_LIST_CONNECTION_CLOSED;
    readonly message: string | null;
}

export interface IOrdersListReceivedAction {
    readonly type: typeof ORDERS_LIST_RECEIVED;
    readonly data: TApiOrders;
}

export type TOrdersListActions =
    | IOrdersListConnectionOpenAction
    | IOrdersListConnectionCloseAction
    | IOrdersListConnectionSuccessAction
    | IOrdersListConnectionFailedAction
    | IOrdersListConnectionClosedAction
    | IOrdersListReceivedAction;
