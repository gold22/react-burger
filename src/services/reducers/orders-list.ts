import {
    ORDERS_LIST_CONNECTION_SUCCESS,
    ORDERS_LIST_CONNECTION_ERROR,
    ORDERS_LIST_CONNECTION_CLOSED,
    ORDERS_LIST_RECEIVED,
} from '../constants/orders-list';
import { TOrders } from '../types';
import { TOrdersListActions } from '../types/actions/orders-list';

export type TOrdersListState = {
    orders: TOrders;
    total: number,
    totalToday: number,
    isConnected: boolean;
    connectionError: string | null;
};

const initialState: TOrdersListState = {
    orders: [],
    total: 0,
    totalToday: 0,
    isConnected: false,
    connectionError: null,
};

export const ordersListReducer = (
    // eslint-disable-next-line @typescript-eslint/default-param-last
    state: TOrdersListState = initialState,
    action: TOrdersListActions,
): TOrdersListState => {
    switch (action.type) {
    case ORDERS_LIST_CONNECTION_SUCCESS: {
        return {
            ...initialState,
            isConnected: true,
        };
    }
    case ORDERS_LIST_CONNECTION_ERROR: {
        return {
            ...initialState,
            // connectionError: action.message,
        };
    }
    case ORDERS_LIST_CONNECTION_CLOSED: {
        return {
            ...initialState,
            // connectionError: action.message,
        };
    }
    case ORDERS_LIST_RECEIVED: {
        return {
            ...state,
            // orders: action.orders,
            // total: action.total,
            // totalToday: action.totalToday,
        };
    }
    default: {
        return state;
    }
    }
};
