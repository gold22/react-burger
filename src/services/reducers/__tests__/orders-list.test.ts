import { ordersListReducer, TOrdersListState } from '../orders-list';
import {
    ORDERS_LIST_CONNECTION_CLOSED,
    ORDERS_LIST_CONNECTION_ERROR,
    ORDERS_LIST_CONNECTION_SUCCESS,
    ORDERS_LIST_RECEIVED,
} from '../../constants/orders-list';
import { IOrdersListReceivedAction } from '../../types/actions/orders-list';
import { apiOrders, orders } from '../../../__fixtures__/orders';

const initialState: TOrdersListState = ordersListReducer(undefined, {} as any);

describe('orders list reducer', () => {
    it('should return the initial state', () => {
        expect(initialState).toMatchSnapshot();
    });

    it('should handle ORDERS_LIST_CONNECTION_SUCCESS', () => {
        expect(ordersListReducer(undefined, { type: ORDERS_LIST_CONNECTION_SUCCESS }))
            .toMatchSnapshot();
    });

    it('should handle ORDERS_LIST_RECEIVED', () => {
        const action: IOrdersListReceivedAction = {
            type: ORDERS_LIST_RECEIVED,
            data: apiOrders,
        };
        const state: TOrdersListState = {
            ...initialState,
            isConnected: true,
        };
        expect(ordersListReducer(state, action))
            .toMatchSnapshot();
    });

    it('should handle ORDERS_LIST_CONNECTION_ERROR', () => {
        expect(ordersListReducer(undefined, { type: ORDERS_LIST_CONNECTION_ERROR, message: null }))
            .toMatchSnapshot();
        expect(ordersListReducer(undefined, { type: ORDERS_LIST_CONNECTION_ERROR, message: 'Connection error' }))
            .toMatchSnapshot();
    });

    it('should handle ORDERS_LIST_CONNECTION_CLOSED', () => {
        const state: TOrdersListState = {
            ...initialState,
            orders,
            total: 9848,
            totalToday: 17,
            isConnected: true,
        };
        expect(ordersListReducer(state, { type: ORDERS_LIST_CONNECTION_CLOSED, message: null }))
            .toMatchSnapshot();
        expect(ordersListReducer(state, { type: ORDERS_LIST_CONNECTION_CLOSED, message: 'Connection close error' }))
            .toMatchSnapshot();
    });
});
