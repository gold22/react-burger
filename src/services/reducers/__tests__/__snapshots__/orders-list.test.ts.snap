// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orders list reducer should handle ORDERS_LIST_CONNECTION_CLOSED 1`] = `
Object {
  "connectionError": null,
  "isConnected": false,
  "orders": Array [
    Object {
      "createdAt": 2022-02-12T16:37:04.209Z,
      "id": "6207e2306d7cd8001b2d5af9",
      "ingredients": Array [
        "60d3b41abdacab0026a733c6",
        "60d3b41abdacab0026a733c8",
        "60d3b41abdacab0026a733c6",
      ],
      "name": "Краторный люминесцентный бургер",
      "number": 9935,
      "status": "created",
      "updatedAt": 2022-02-12T16:37:04.448Z,
    },
    Object {
      "createdAt": 2022-02-12T16:35:50.383Z,
      "id": "6207e1e66d7cd8001b2d5af7",
      "ingredients": Array [
        "60d3b41abdacab0026a733c7",
        "60d3b41abdacab0026a733cc",
        "60d3b41abdacab0026a733c7",
      ],
      "name": "Spicy флюоресцентный бургер",
      "number": 9934,
      "status": "pending",
      "updatedAt": 2022-02-12T16:35:50.614Z,
    },
    Object {
      "createdAt": 2022-02-12T15:47:31.913Z,
      "id": "6207d6936d7cd8001b2d5a91",
      "ingredients": Array [
        "60d3b41abdacab0026a733c6",
        "60d3b41abdacab0026a733ce",
        "60d3b41abdacab0026a733c6",
      ],
      "name": "Традиционный-галактический краторный бургер",
      "number": 9932,
      "status": "done",
      "updatedAt": 2022-02-12T15:47:32.195Z,
    },
  ],
  "total": 9848,
  "totalToday": 17,
}
`;

exports[`orders list reducer should handle ORDERS_LIST_CONNECTION_CLOSED 2`] = `
Object {
  "connectionError": "Connection closed error",
  "isConnected": false,
  "orders": Array [],
  "total": 0,
  "totalToday": 0,
}
`;

exports[`orders list reducer should handle ORDERS_LIST_CONNECTION_ERROR 1`] = `
Object {
  "connectionError": null,
  "isConnected": false,
  "orders": Array [],
  "total": 0,
  "totalToday": 0,
}
`;

exports[`orders list reducer should handle ORDERS_LIST_CONNECTION_ERROR 2`] = `
Object {
  "connectionError": "Connection error",
  "isConnected": false,
  "orders": Array [],
  "total": 0,
  "totalToday": 0,
}
`;

exports[`orders list reducer should handle ORDERS_LIST_CONNECTION_SUCCESS 1`] = `
Object {
  "connectionError": null,
  "isConnected": true,
  "orders": Array [],
  "total": 0,
  "totalToday": 0,
}
`;

exports[`orders list reducer should handle ORDERS_LIST_RECEIVED 1`] = `
Object {
  "connectionError": null,
  "isConnected": true,
  "orders": Array [
    Object {
      "createdAt": 2022-02-12T15:47:31.913Z,
      "id": "6207d6936d7cd8001b2d5a91",
      "ingredients": Array [
        "60d3b41abdacab0026a733c6",
        "60d3b41abdacab0026a733ce",
        "60d3b41abdacab0026a733c6",
      ],
      "name": "Традиционный-галактический краторный бургер",
      "number": 9932,
      "status": "done",
      "updatedAt": 2022-02-12T15:47:32.195Z,
    },
    Object {
      "createdAt": 2022-02-12T16:35:50.383Z,
      "id": "6207e1e66d7cd8001b2d5af7",
      "ingredients": Array [
        "60d3b41abdacab0026a733c7",
        "60d3b41abdacab0026a733cc",
        "60d3b41abdacab0026a733c7",
      ],
      "name": "Spicy флюоресцентный бургер",
      "number": 9934,
      "status": "pending",
      "updatedAt": 2022-02-12T16:35:50.614Z,
    },
    Object {
      "createdAt": 2022-02-12T16:37:04.209Z,
      "id": "6207e2306d7cd8001b2d5af9",
      "ingredients": Array [
        "60d3b41abdacab0026a733c6",
        "60d3b41abdacab0026a733c8",
        "60d3b41abdacab0026a733c6",
      ],
      "name": "Краторный люминесцентный бургер",
      "number": 9935,
      "status": "created",
      "updatedAt": 2022-02-12T16:37:04.448Z,
    },
  ],
  "total": 9848,
  "totalToday": 17,
}
`;

exports[`orders list reducer should return the initial state 1`] = `
Object {
  "connectionError": null,
  "isConnected": false,
  "orders": Array [],
  "total": 0,
  "totalToday": 0,
}
`;
